libfwupdplugin/fu-bluez-device.c

```diff
@@ -293,59 +291,151 @@ fu_bluez_device_get_ble_string_property(const gchar *obj_path,
 	return g_variant_dup_string(val, NULL);
 }
 
+static gchar *
+fu_bluez_device_get_interface_uuid(FuBluezDevice *self,
+				   GDBusObject *obj,
+				   const gchar *obj_path,
+				   const gchar *iface_name)
+{
+	g_autoptr(GDBusInterface) iface = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autofree gchar *obj_uuid = NULL;
+
+	iface = g_dbus_object_get_interface(obj, iface_name);
+	if (iface == NULL)
+		return NULL;
+	obj_uuid =
+	    fu_bluez_device_get_ble_string_property(obj_path, iface_name, "UUID", &error_local);
+	if (obj_uuid == NULL)
+		g_debug("failed to get %s property: %s", iface_name, error_local->message);
+
+	return g_steal_pointer(&obj_uuid);
+}
+
+/*
+ * Populates the {uuid_helper : object_path} entry of a device for its
+ * characteristic.
+ */
+static gboolean
+fu_bluez_device_add_characteristic_uuid(FuBluezDevice *self,
+					GDBusObject *obj,
+					const gchar *obj_path,
+					const gchar *iface_name)
+{
+	g_autofree gchar *obj_uuid =
+	    fu_bluez_device_get_interface_uuid(self, obj, obj_path, iface_name);
+
+	if (obj_uuid == NULL)
+		return FALSE;
+
+	fu_bluez_device_add_uuid_path(self, obj_uuid, obj_path);
+	return TRUE;
+}
+
+static gboolean
+fu_bluez_device_add_instance_by_service_uuid(FuBluezDevice *self,
+					     GDBusObject *obj,
+					     const gchar *obj_path,
+					     const gchar *iface_name)
+{
+	g_autoptr(GError) error_local = NULL;
+	g_autofree gchar *obj_uuid =
+	    fu_bluez_device_get_interface_uuid(self, obj, obj_path, iface_name);
+
+	if (obj_uuid == NULL)
+		return FALSE;
+
+	/* Register device by service UUID. */
+	fu_device_add_instance_str(FU_DEVICE(self), "GATT", obj_uuid);
+	if (!fu_device_build_instance_id_full(FU_DEVICE(self),
+					      FU_DEVICE_INSTANCE_FLAG_VISIBLE |
+						  FU_DEVICE_INSTANCE_FLAG_QUIRKS,
```

Comment:
I think lose the `FU_DEVICE_INSTANCE_FLAG_VISIBLE` and then we can still match it in quirks for "interesting device" rather than showing them all to the user as potential matches for firmware.

Reply:
hm... I did a try and have:
```
11:59:10.360 FuEngine             /org/bluez/hci0/dev_00_02_5B_00_FF_01 added FuBluezDevice:
  Name:                 CCC5171
  Flags:                none
  BatteryLevel:         98
...
  Guid[quirk]:          21608f80-6c9c-5cd1-91ed-0fbf18e31958 <- BLUETOOTH\GATT_00001100-d102-11e1-9b23-00025b00a5a5
...
  PhysicalId:           /org/bluez/hci0
  LogicalId:            00:02:5B:00:FF:01
  BackendId:            /org/bluez/hci0/dev_00_02_5B_00_FF_01
...
11:59:11.061 FuEngine             no GUIDs for device CCC5171 [9f455188df0fd268b4b349575f714570967c3815]
```

The quirk file:
```
[BLUETOOTH\GATT_00001100-d102-11e1-9b23-00025b00a5a5]
Plugin = audio_s5gen2
ProxyGType = FuQcS5gen2BleDevice
```

Works well with `FU_DEVICE_INSTANCE_FLAG_VISIBLE` flag.
Maybe I missed something else?
I'm trying to register the device, which is working via BLE proxy implementing the Interface for communication.

Reply:
> no GUIDs for device CCC5171

That's actually success! We don't want to use any of the GATT UUIDs for firmware matching. I think you can fix this by adding a GUID (that is used for matching firmware) manually in the quirk file -- maybe even `USB\VID_0A12&PID_4007`? I think we need to work out how to get a proper VID and PID for the FuQcS5gen2BleDevice device really -- so we know what firmware to suggest.

Reply:
Can you dump what you see for FuQcS5gen2BleDevice please?